Support bfp8 inputs in convert_to_chw, improve perf

pavlepopovic opened the pull request:
### Ticket
#20185: 

### Problem description
in ticket

### What's changed
Suport bfp8 inputs in convert_to_chw, improve it's perf (it was doing lots of unnecessary inits/uninits)
Shallow unet chw op: from 51k cycles down to 21k
Customer model: from 72k cycles down to 28k

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [x] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [x] New/Existing tests provide coverage for changes

Will update device/e2e perf following device/e2e pipelines

Commit:
#20185: Support bfp8 inputs in convert_to_chw, improve perf

pavlejosipovic commented on ttnn/cpp/ttnn/operations/experimental/cnn/convert_to_chw/device/kernels/convert_to_chw.cpp:
```diff
@@ -45,8 +39,11 @@ void MAIN {
+    transpose_wh_init_short(cb_in);
+    pack_untilize_dst_init_short<ONE_TILE, 1, false, false, 32>(cb_out);
     for (uint32_t idx = 0; idx < total_tiles; idx++) {
         transpose(cb_in, cb_transpose_in);
```
seems like we should be doing 8 by 8 tiles in one go?

esmalTT replied:
yeah would give another nice boost

pavlepopovic replied:
Yup, tried doing that and got stuck in hangs which I don't have time to debug atm, will revisit sometime :)

esmalTT replied:
same thing happened to me when i wrote this initially 🤣

pavlejosipovic replied:
https://github.com/tenstorrent/tt-metal/tree/pjosipovic/convert_to_chw
Here is the bracnh on top of your commit.

"Add batching to compute and writer kernels.
Add double buffering between compute and writer kernels.
Move writer kernel to ncrisk.
2x performance improvement."

esmalTT reviewed (approved):
This is great, thanks for the perf! LGTM assuming model pipelines are good 👍

bbradelTT commented on tests/ttnn/unit_tests/operations/test_convert_to_chw.py:
```diff
@@ -38,15 +46,19 @@ def test_convert_to_chw(device, C, HW, core_grid):
-    assert_with_pcc(expected, ttnn.to_torch(actual), 1.0)
+    expected_pcc = 1.0
+    if input_data_type == ttnn.bfloat8_b:
+        expected_pcc = 0.9999  # bfloat8_b can't be exatcly compared to torch bfloat16
```
Should we look at asserting equality directly for non bfloat8_b?

pavlepopovic replied:
For this op, yes. It’s not doing any compute, just data movement. If its bf16 then ti should be the same

pavlepopovic replied:
Oh you mean like skip pcc check and just compare == ?

pavlepopovic replied:
added assert_equal if we're dealing with bf16

pavlepopovic commented:
Updated perf targets:
- e2e single/multi chip https://github.com/tenstorrent/tt-metal/actions/runs/14267581009/job/39993418261
- device https://github.com/tenstorrent/tt-metal/actions/runs/14267575533/job/39993380570

all post commits (failure in an unrelelated test, seems like a machine instability):
https://github.com/tenstorrent/tt-metal/actions/runs/14267571780/job/39993380177

Commit:
Do batched processing for convert_to_chw

Add batching to compute and writer kernels.
Add double buffering between compute and writer kernels.
Move writer kernel to ncrisk.
2x performance improvement.

Commit:
#0: Use assert_equal

omilyutin-tt commented on ttnn/cpp/ttnn/operations/experimental/cnn/convert_to_chw/device/convert_to_chw_program_factory.cpp:
```diff
@@ -63,34 +65,35 @@ tt::tt_metal::operation::ProgramWithCallbacks multi_core_convert_to_chw(
     std::vector<uint32_t> reader_compile_time_args = {cb_in_id};
     std::vector<uint32_t> writer_compile_time_args = {cb_in_transpose_id, cb_out_id, C};
-    std::vector<uint32_t> compute_compile_time_args = {cb_in_id, cb_in_transpose_id, cb_out_id};
+    std::vector<uint32_t> compute_compile_time_args = {cb_in_id, cb_in_transpose_id};
```
This code now reads 

```
tt::tt_metal::WriterDataMovementConfig(reader_compile_time_args);
```

Is this correct? Can we rename reader/writer compile time args accordingly?

pavlepopovic replied:
fixed ty :)

omilyutin-tt reviewed (approved):
Approving to avoid blocking

Commit:
#0: Rebase + PR changes

yechanjun commented:
How the `constexpr int BATCH_SIZE = 8;` is determined?

pavlepopovic commented:
Dest register has space for 16 tiles in total, since we're using double buffering between dest/packer that means we have 8 tiles in total to use at a time. Hence the BATCH_SIZE = 8.
Without this, math will produce a single tile, and packer will pack it, constantly switching between the two.
Now, math will produce 8 tiles, and give it to packer, reducing number of switches

pavlepopovic commented:
math acquires them with tile_regs_acquire(), releases with tile_regs_commit()
pack acquires them with tile_regs_wait(), releases with tile_regs_release()
Let's take a look at tile_regs_acguire which translates to llk_math_wait_for_dest_available()
```
inline void llk_math_wait_for_dest_available() {
    WAYPOINT("MWDW");
    _llk_math_wait_for_dest_available_<DST_SYNC_MODE>();
    WAYPOINT("MWDD");
}
```
The key param here is tempalte DST_SYNC_MODE, and further in the llks it is defined as following:
```
enum DstSync
{
    SyncHalf   = 0,
    SyncFull   = 1,
    SyncTile16 = 2,
    SyncTile2  = 3,
};
```
It is the resposibility of the OP to define which sync mode to use, and in this kernel, it is DstSync::SyncHalf -> meaning upon calling acquire/release, only half of the dest is in question. When math finishes, packer can start packing on that half, while math picks up the other half.

Btw you can check which one got generated for the kernel in the built folder - an example: built/tt-metal-cache/f49e23dd32/4097/kernels/convert_to_chw/13416792310708376122/chlkc_dst_sync_mode.h

Does that make things more understandable?  😄

yechanjun commented:
Amazing, You are my hero ! It was so helpful to understand the concept of double-buffering :)

pavlepopovic commented:
Glad to have helped :)
